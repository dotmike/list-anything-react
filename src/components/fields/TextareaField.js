import React from "react";
import PropTypes from "prop-types";
import { Form } from "semantic-ui-react";

import InlineError from "../messages/InlineError";

const onClick = e => {
  e.preventDefault();
  e.stopPropagation();
};

const TextareaField = ({
  value,
  editable,
  showFieldDescription,
  error,
  field,
  onChange
}) => {
  if (editable)
    return (
      <Form.Field error={!!error}>
        <Form.TextArea
          autoHeight
          style={{ maxHeight: "200px", minHeight: "75px" }}
          label={field.description}
          value={value}
          onChange={onChange}
          name={field._id}
          placeholder={field.description}
        />
        {error && <InlineError text={error} />}
      </Form.Field>
    );
  return (
    <Form.Field readOnly>
      <Form.TextArea
        autoHeight
        style={{
          maxHeight: showFieldDescription ? "200px" : "150px",
          border: "none",
          minWidth: "100%",
          outline: "none"
        }}
        label={showFieldDescription && field.description}
        value={value}
        name={field._id}
        placeholder={field.description}
        onClick={onClick}
      />
    </Form.Field>
  );
};

TextareaField.propTypes = {
  // ownProps
  value: PropTypes.string.isRequired,
  editable: PropTypes.bool,
  showFieldDescription: PropTypes.bool,
  error: PropTypes.string,
  field: PropTypes.shape({
    _id: PropTypes.string,
    description: PropTypes.string
  }),
  onChange: PropTypes.func
};

TextareaField.defaultProps = {
  editable: false,
  showFieldDescription: true,
  error: "",
  field: {},
  onChange: () => {}
};

export default TextareaField;
