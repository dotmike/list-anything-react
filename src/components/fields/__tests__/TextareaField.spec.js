import TextareaField from "../TextareaField";

const testData = {
  value: "testando...",
  field: { _id: "1", description: "Testando:" },
  newValue: "test"
};

const setup = (propOverrides = {}) => {
  const props = {
    value: testData.value,
    error: "",
    editable: true,
    showFieldDescription: true,
    field: testData.field,
    onChange: jest.fn(),
    ...propOverrides
  };

  return {
    props,
    wrapperShallow: wrapperShallow(TextareaField, props)
  };
};

describe("TextareaField", () => {
  describe("when `editable` is true", () => {
    it("renders correctly", () => {
      const { wrapperShallow: wrapper } = setup({
        editable: true
      });
      expect(wrapper()).toMatchSnapshot();
    });

    it("renders correctly when `error` is passed", () => {
      const { wrapperShallow: wrapper } = setup({
        editable: true,
        error: "Something went wrong..."
      });
      expect(wrapper()).toMatchSnapshot();
    });

    it("calls `onChange` when changing the input value", () => {
      const { wrapperShallow: wrapper, props } = setup({
        editable: true
      });

      wrapper()
        .find("FormTextArea")
        .simulate("change", { target: { value: testData.newValue } });
      expect(props.onChange).toHaveBeenCalled();
    });
  });

  describe("when `editable` is false", () => {
    it("renders correctly when `showFieldDescription` is true", () => {
      const { wrapperShallow: wrapper } = setup({
        editable: false,
        showFieldDescription: true
      });
      expect(wrapper()).toMatchSnapshot();
    });

    it("renders correctly when `showFieldDescription` is false", () => {
      const { wrapperShallow: wrapper } = setup({
        editable: false,
        showFieldDescription: false
      });
      expect(wrapper()).toMatchSnapshot();
    });
  });
});
