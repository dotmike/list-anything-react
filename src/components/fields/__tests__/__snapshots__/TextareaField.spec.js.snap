// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextareaField when \`editable\` is false renders correctly when \`showFieldDescription\` is false 1`] = `
<FormField
  readOnly={true}
>
  <FormTextArea
    as={[Function]}
    autoHeight={true}
    control={[Function]}
    label={false}
    name="1"
    onClick={[Function]}
    style={
      Object {
        "border": "none",
        "maxHeight": "150px",
        "minWidth": "100%",
        "outline": "none",
      }
    }
    value="testando..."
  />
</FormField>
`;

exports[`TextareaField when \`editable\` is false renders correctly when \`showFieldDescription\` is true 1`] = `
<FormField
  readOnly={true}
>
  <FormTextArea
    as={[Function]}
    autoHeight={true}
    control={[Function]}
    label="Testando:"
    name="1"
    onClick={[Function]}
    style={
      Object {
        "border": "none",
        "maxHeight": "200px",
        "minWidth": "100%",
        "outline": "none",
      }
    }
    value="testando..."
  />
</FormField>
`;

exports[`TextareaField when \`editable\` is true renders correctly 1`] = `
<FormField
  error={false}
>
  <FormTextArea
    as={[Function]}
    autoHeight={true}
    control={[Function]}
    label="Testando:"
    name="1"
    onChange={[Function]}
    placeholder="Testando:"
    style={
      Object {
        "maxHeight": "200px",
        "minHeight": "75px",
      }
    }
    value="testando..."
  />
</FormField>
`;

exports[`TextareaField when \`editable\` is true renders correctly when \`error\` is passed 1`] = `
<FormField
  error={true}
>
  <FormTextArea
    as={[Function]}
    autoHeight={true}
    control={[Function]}
    label="Testando:"
    name="1"
    onChange={[Function]}
    placeholder="Testando:"
    style={
      Object {
        "maxHeight": "200px",
        "minHeight": "75px",
      }
    }
    value="testando..."
  />
  <InlineError
    text="Something went wrong..."
  />
</FormField>
`;
